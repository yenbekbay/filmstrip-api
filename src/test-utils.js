/* @flow */

import _ from 'lodash/fp';
import type from 'type-detect';

const childTypesForArray = _.memoize(_.flow(_.map(type), _.uniq));

const modelFromObject = (obj: ?Object) =>
  _.mapValues(
    _.cond([
      [
        _.overEvery([
          _.flow(type, _.eq('Array')),
          _.flow(childTypesForArray, _.head, _.eq('Object')),
        ]),
        (val: Array<any>) => [_.flow(_.head, modelFromObject)(val)],
      ],
      [
        _.overEvery([
          _.flow(type, _.eq('Array')),
          _.flow(childTypesForArray, _.size, _.eq(1)),
        ]),
        (val: Array<any>) =>
          `Array<${_.flow(childTypesForArray, _.head)(val)}>`,
      ],
      [_.flow(type, _.includes(_, ['Object', 'Array'])), modelFromObject],
      [_.stubTrue, type],
    ]),
    obj,
  );

export {modelFromObject}; // eslint-disable-line import/prefer-default-export
