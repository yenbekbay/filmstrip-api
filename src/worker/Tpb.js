/* @flow */

import _ from 'lodash/fp';
import bytes from 'bytes';
import PirateBay from 'thepiratebay';
import retry from 'async-retry';

import { torrentQualityTest } from './utils';
import type { Torrent } from '../types';

const HD_MOVIES_CATEGORY = 207;
const MAX_TORRENTS = 5;
const blacklistedUploaders = [
  'jXTENZ8',
];

const yearFromNameRegex = /(?:\(|\s+|\.)(\d{4})(?:\)|\s+|\.)/;
const titleFromNameRegex = /(.+)(?:\(|\s+|\.)(?:\d{4})(?:\)|\s+|\.)/;

class Tpb {
  getTorrentsForMovie = async (
    query: { title: string, year: number },
  ): Promise<Array<Torrent>> => retry(
    async (bail: (err: Error) => void) => {
      const { title, year } = query;
      const sanitizedMovieTitle = title.replace('\'s', '');

      try {
        const results = await PirateBay.search(
          `${sanitizedMovieTitle} ${year} 1080p|720p -HC`,
          {
            category: HD_MOVIES_CATEGORY,
            orderBy: 'seeds',
            sortBy: 'desc',
          },
        );

        return _.flow(
          _.map(({ name, size, seeders, leechers, magnetLink }: {
            name: string,
            size: string,
            seeders: string,
            leechers: string,
            magnetLink: string,
          }): Torrent => ({
            source: 'The Pirate Bay',
            name,
            size: bytes(size.replace(/\s+/, '').replace('i', '')),
            seeds: parseInt(seeders, 10),
            peers: parseInt(leechers, 10),
            quality: _.cond([
              [_.includes('720p'), _.constant('720p')],
              [_.includes('1080p'), _.constant('1080p')],
              [_.stubTrue, _.constant(null)],
            ])(name),
            magnetLink,
          })),
          _.filter(_.overEvery([
            torrentQualityTest,
            ({ seeds }: Torrent) => seeds > 10,
          ])),
          _.slice(0, MAX_TORRENTS),
        )(results);
      } catch (err) {
        if (_.includes('ECONNRESET', err.message)) {
          throw err;
        }

        return bail(err);
      }
    },
    { retries: 5 },
  );

  getTopMovies = async () => {
    const currentYear = new Date().getFullYear();

    try {
      const results = await PirateBay.topTorrents(HD_MOVIES_CATEGORY);

      return _.flow(
        _.map(({ name, uploader }: Object) => ({
          name,
          uploader,
          year: parseInt(_.nth(1, name.match(yearFromNameRegex)), 10),
        })),
        _.filter(
          ({ name, year, uploader }: Object) =>
            year && year >= (currentYear - 1) &&
            !_.includes('HC', name) &&
            !_.includes(uploader, blacklistedUploaders),
        ),
        _.map(({ name, year }: Object) => {
          const title = _.flow(
            _.nth(1),
            _.replace(/\./g, ' '),
            _.replace(/\s+/g, ' '),
            _.split('/'),
            _.head,
            _.trim,
          )(name.match(titleFromNameRegex));

          return { title, year };
        }),
        _.uniqBy('title'),
      )(results);
    } catch (err) {
      if (_.includes('ECONNRESET', err.message)) {
        return this.getTopMovies();
      }

      throw err;
    }
  };
}

export default Tpb;
