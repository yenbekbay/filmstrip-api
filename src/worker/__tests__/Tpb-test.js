/* @flow */

import { modelFromObject } from '../../test-utils';
import Tpb from '../Tpb';

const movieTitle = 'Pete\'s Dragon';

describe('getTpbTorrents', () => {
  let tpb: Tpb;

  beforeAll(() => {
    tpb = new Tpb();
  });

  it('fetches a list of torrents for a given query', async () => {
    const torrents = await tpb.getTorrentsForMovie(movieTitle);
    expect(modelFromObject({ torrents })).toMatchSnapshot();
  });

  it('fetches a list of top movies', async () => {
    const movies = await tpb.getTopMovies();
    expect(modelFromObject({ movies })).toMatchSnapshot();
  });
});
