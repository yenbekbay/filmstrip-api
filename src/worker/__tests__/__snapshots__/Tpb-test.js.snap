exports[`getTpbTorrents fetches a list of top movies 1`] = `
Object {
  "movies": Array [
    Object {
      "title": "string",
      "year": "number",
    },
  ],
}
`;

exports[`getTpbTorrents fetches a list of torrents for a given query 1`] = `
Object {
  "torrents": Array [
    Object {
      "magnetLink": "string",
      "name": "string",
      "peers": "number",
      "quality": "string",
      "seeds": "number",
      "size": "number",
      "source": "string",
    },
  ],
}
`;
